src/backend/x11/mod.rs

```diff
@@ -0,0 +1,921 @@
+                            (configure_notify.width, configure_notify.height).into();
+
+                        if configure_notify_size != previous_size {
+                            // Intentionally drop the lock on the size mutex incase a user
+                            // requests a resize or does something which causes a resize
+                            // inside the callback.
+                            {
+                                *window.size.lock().unwrap() = configure_notify_size;
+                            }
+
+                            (callback)(X11Event::Resized(configure_notify_size), &mut event_window);
+                            let _ = resize.send(configure_notify_size);
+                        }
+                    }
+                }
+
+                x11::Event::ClientMessage(client_message) => {
+                    if client_message.data.as_data32()[0] == window.atoms.WM_DELETE_WINDOW // Destroy the window?
+                            && client_message.window == window.id
+                    // Same window
+                    {
+                        (callback)(X11Event::CloseRequested, &mut event_window);
+                    }
+                }
+
+                x11::Event::Expose(expose) => {
+                    if expose.window == window.id && expose.count == 0 {
+                        (callback)(X11Event::Refresh, &mut event_window);
+                    }
+                }
+
+                x11::Event::Error(e) => {
+                    error!(log, "X11 error: {:?}", e);
+                }
+
+                _ => (),
+            }
+
+            // Flush the connection so changes to the window state during callbacks can be emitted.
+            let _ = connection.flush();
+        })
+    }
+
+    fn register(&mut self, poll: &mut Poll, token_factory: &mut TokenFactory) -> io::Result<()> {
+        self.source.register(poll, token_factory)
+    }
+
+    fn reregister(&mut self, poll: &mut Poll, token_factory: &mut TokenFactory) -> io::Result<()> {
+        self.source.reregister(poll, token_factory)
+    }
+
+    fn unregister(&mut self, poll: &mut Poll) -> io::Result<()> {
+        self.source.unregister(poll)
+    }
+}
+
+fn check_for_extensions(connection: &RustConnection, logger: &Logger) -> Result<(), X11Error> {
+    // Xfixes
+    {
+        if connection
+            .extension_information(xfixes::X11_EXTENSION_NAME)?
+            .is_none()
+        {
+            error!(logger, "Xfixes extension not found");
+            return Err(MissingExtensionError::NotFound {
+                name: xfixes::X11_EXTENSION_NAME,
+                major: XFIXES_MAJOR_VERSION,
+                minor: XFIXES_MINOR_VERSION,
+            }
+            .into());
+        }
+
+        let version = connection
+            .xfixes_query_version(XFIXES_MAJOR_VERSION, XFIXES_MINOR_VERSION)?
+            .reply()?;
+
+        if version.major_version < XFIXES_MAJOR_VERSION {
```

Comment:
This ignores `XFIXES_MINOR_VERSION`. Sure, that constant is currently 0, but... yeah.

I would recommend a `compare_versions` helper function:
```rust
fn compare_versions(actual: (u32, u32), expected: (u32, u32)) -> bool {
    actual.0 > expected.0 || (actual.0 == expected.0 && actual.1 >= expected.1)
}
```
Or perhaps even:
````rust
fn compare_versions(name: &'static str, actual: (u32, u32), expected: (u32, u32)) -> Result<(), MissingExtensionError> {
    if actual.0 > expected.0 || (actual.0 == expected.0 && actual.1 >= expected.1) {
        Ok(())
    } else {
        Err(MissingExtensionError::WrongVersion { ..... })
    }
}
````